priority: HTTP/2 doesn't have stream limits (#1344)
httpwg/http-extensions: draft-ietf-httpbis-priority.md

The following example shows a request for a CSS file with the urgency set to
`0`:

> ~~~ http-sf-dict

Reviewer: Lines 240, 279, 497, and 507 are all lists of header fields.  Why is this one a dict and the others lists?

Author: oh this is funny, I assumed this was some automagic tool or process that applied the changes but I think it is actually a hangover from an experiment around structured fields. I presume then that Istarted the PR in a dirty repo. I'll just take these out because I think the experiment took us in a different direction.

HTTP/3 the error is of type H3_FRAME_ERROR.

A client MAY send a PRIORITY_UPDATE frame before the stream that it references

> is open (except for HTTP/2 push streams; see {{h2-update-frame}}. Furthermore,

Reviewer 2: ```suggestion
is open (except for HTTP/2 push streams; see {{h2-update-frame}}). Furthermore,
```

Priority Field Value:
: The priority update value in ASCII text, encoded using Structured Fields.

The Prioritized Stream ID SHOULD be a value within the stream concurrency

> window. A server SHOULD NOT treat a Prioritized Stream ID beyond this window as

Reviewer: This doesn't appear to be a term defined in RFC 7540.  I think I'd try to frame this in terms of the number of active streams ("open" state or either "half-closed" state) and the value of SETTINGS_MAX_CONCURRENT_STREAMS, both of which are defined for HTTP/2.

Author: Thanks for calling me up on this, its exactly the type of review that's needed.

Priority Field Value:
: The priority update value in ASCII text, encoded using Structured Fields.

The Prioritized Stream ID SHOULD be a value within the limit of concurrent
active streams, which is determined by the SETTINGS_MAX_CONCURRENT_STREAMS

> paramater and streams in the "open" or "half-closed" state; see Section 5.1.2 of

Reviewer: ```suggestion
parameter and streams in the "open" or "half-closed" state; see Section 5.1.2 of
```

Priority Field Value:
: The priority update value in ASCII text, encoded using Structured Fields.

When the PRIORITY_UPDATE frame applies to a request stream, a client SHOULD
provide a Prioritized Stream ID that refers to a stream in the "open",

> "half-closed (local)", or "idle" state. Frames where the Prioritized Stream ID
> refers to a stream in the "half-closed (remote)" or "closed" state can be
> discarded by the server upon receipt. The number of streams which have been

Reviewer: ```suggestion
"half-closed (local)", or "idle" state. The server can discard frames where the
Prioritized Stream ID refers to a stream in the "half-closed (local)" or "closed"
state can be discarded upon receipt. The number of streams which have been
```
"Half-closed (local)" on the server is "half-closed (remote)" on the client.  The switch of perspective can be signaled by making the server the subject of the sentence.

prioritized but remain in the "idle" state plus the number of active streams

> (those in the "open" or "half-closed" states; see section 5.1.2 of RFC 7540)

Reviewer: ```suggestion
(those in the "open" or either "half-closed" state; see section 5.1.2 of RFC 7540)
```
Needs rewrap.

MUST NOT exceed the value of the SETTINGS_MAX_CONCURRENT_STREAMS parameter. A
server that receives such a PRIORITY_UPDATE MUST respond with a connection error
of type PROTOCOL_ERROR.

When the PRIORITY_UPDATE frame applies to a push stream, a client SHOULD provide

> a Prioritized Stream ID that refers to a stream in the "reserved (remote)",
> "half-closed (local)", or "idle" state. Frames where the Prioritized Stream ID

Reviewer: ```suggestion
a Prioritized Stream ID that refers to a stream in the "reserved (remote)" or
"half-closed (local)" state. Frames where the Prioritized Stream ID
```
You can't prioritize a push in the idle state, because it hasn't been promised yet; you have no idea what it is.  Further, the reason there's not a limit on prioritizing pushed streams is because it's bound by the number of promises the server has issued.

It's not guaranteed detectable (because the PRIORITY_UPDATE could cross the PUSH_PROMISE on the network), but receiving prioritization for a push which is still "idle" should probably be an error.  That implies the client MUST NOT send PRIORITY_UPDATE for a server-initiated idle stream.

Author: I added that advice to the text, thanks.

refers to a stream in the "half-closed (remote)" or "closed" state can be

> discarded by the server upon receipt. The number of streams which have been
> prioritized but remain in the "idle" state plus the number of active streams
> (those in the "reserved" or "half-closed" states; see section 5.1.2 of RFC 7540)
> MUST NOT exceed the value of the SETTINGS_MAX_CONCURRENT_STREAMS parameter. A
> server that receives such a PRIORITY_UPDATE MUST respond with a connection error
> of type PROTOCOL_ERROR.

Reviewer: ```suggestion
discarded by the server upon receipt.
```
This isn't needed for pushed streams; the server controls the number of active pushes, and the client can't prioritize something the server hasn't pushed.

Priority Field Value:
: The priority update value in ASCII text, encoded using Structured Fields.

When the PRIORITY_UPDATE frame applies to a request stream, clients SHOULD
provide a Prioritized Stream ID that refers to a stream in the "open",
"half-closed (local)", or "idle" state. Servers can discard frames where the
Prioritized Stream ID refers to a stream in the "half-closed (local)" or
"closed" state. The number of streams which have been prioritized but remain in
the "idle" state plus the number of active streams (those in the "open" or
either "half-closed" state; see section 5.1.2 of {{RFC7540}}) MUST NOT exceed
the value of the SETTINGS_MAX_CONCURRENT_STREAMS parameter. Servers that receive
such a PRIORITY_UPDATE MUST respond with a connection error of type
PROTOCOL_ERROR.

When the PRIORITY_UPDATE frame applies to a push stream, clients SHOULD provide
a Prioritized Stream ID that refers to a stream in the "reserved (remote)" or
"half-closed (local)" state. Servers can discard frames where the Prioritized
Stream ID refers to a stream in the "half-closed (remote)" or "closed" state.

> Client MUST NOT provide a Prioritized Stream ID that refers to a stream in the

Reviewer 2: ```suggestion
Clients MUST NOT provide a Prioritized Stream ID that refers to a stream in the
```

> "idle" state. Servers that receive such a PRIORITY_UPDATE MUST respond with a
> connection error of type PROTOCOL_ERROR.

Reviewer 2: ```suggestion
"idle" state. Servers that receive a PRIORITY_UPDATE for a stream in the "idle" state MUST
respond with a connection error of type PROTOCOL_ERROR.
```
